### PR TITLE
libav changes

### PR DESCRIPTION
**Pull request by naushir**

### CODE REVIEW

**Comment by naushir** [comment]
@davidplowman can you cast your eyes on this please?  Hopefully should give parity in the command line args between PI 4 and 5 for rpicam-vid.

File: encoder/libav_encoder.cpp [hunk 1]
```diff
@@ -173,17 +173,29 @@ void LibAvEncoder::initVideoCodec(VideoOptions const *options, StreamInfo const
 	// Apply general options.
 	encoderOptionsGeneral(options, codec_ctx_[Video]);
 
-	const char *format;
-	if (options_->output.empty() && options->libav_format.empty())
-		format = "h264";
-	else if (options->libav_format.empty())
-		format = nullptr;
+	const std::string tcp { "tcp://" };
+	const std::string udp { "udp://" };
```
Comment (davidplowman) [inline encoder/libav_encoder.cpp:177]: I'm assuming we don't care about case here!

File: encoder/libav_encoder.cpp [hunk 2]
```diff
@@ -173,17 +173,29 @@ void LibAvEncoder::initVideoCodec(VideoOptions const *options, StreamInfo const
...
+
+	// Setup an appropriate stream/container format.
+	const char *format = nullptr;
+	if (options->libav_format.empty())
+	{
+		if (options->output.empty() ||
+			options->output.find(tcp.c_str(), 0, tcp.length()) != std::string::npos ||
+			options->output.find(udp.c_str(), 0, udp.length()) != std::string::npos)
```
Comment (davidplowman) [inline encoder/libav_encoder.cpp]: This is slightly awkward to read, I find. Are we using C++20? Apparently that has a `starts_with`...

File: encoder/libav_encoder.cpp [hunk 3]
```diff
@@ -180,9 +180,8 @@ void LibAvEncoder::initVideoCodec(VideoOptions const *options, StreamInfo const
 	const char *format = nullptr;
 	if (options->libav_format.empty())
 	{
-		if (options->output.empty() ||
-			options->output.find(tcp.c_str(), 0, tcp.length()) != std::string::npos ||
-			options->output.find(udp.c_str(), 0, udp.length()) != std::string::npos)
+		if (output_file_.empty() || output_file_.find(tcp.c_str(), 0, tcp.length()) != std::string::npos ||
+			output_file_.find(udp.c_str(), 0, udp.length()) != std::string::npos)
```
Comment (davidplowman) [inline encoder/libav_encoder.cpp:184]: Did the style checker request this change? I think I preferred the original!

File: encoder/libav_encoder.cpp [hunk 4]
```diff
@@ -192,8 +191,17 @@ void LibAvEncoder::initVideoCodec(VideoOptions const *options, StreamInfo const
 	}
 	else
 		format = options->libav_format.c_str();
+
+	// Legacy handling of the --listen parameter.  If missing from the url string, add "?listen=1" to the end.
+	if (options->listen && output_file_.find(tcp.c_str(), 0, tcp.length()) != std::string::npos)
+	{
+		const std::string listen { "?listen=1" };
+		if (output_file_.find(listen, output_file_.length() - listen.length()) == std::string::npos)
```
Comment (davidplowman) [inline encoder/libav_encoder.cpp:199]: Just slightly wondering whether it's necessary to constrain the length of the search here, wouldn't that happen by itself?

**Comment by davidplowman** [comment]
My minor whinging notwithstanding, I'm happy to merge...!

File: encoder/libav_encoder.cpp (hunk 1 above)
Comment (naushir) [reply encoder/libav_encoder.cpp:177]: I think this is case sensitive in libav.

File: encoder/libav_encoder.cpp (hunk 2 above)
Comment (naushir) [reply encoder/libav_encoder.cpp]: Unfortunately we are using C++17, but starts_with is exactly what I want :(

File: encoder/libav_encoder.cpp (hunk 3 above)
Comment (naushir) [reply encoder/libav_encoder.cpp:184]: Yes.  I prefer it as the original as well, I'll revert it.

File: encoder/libav_encoder.cpp (hunk 4 above)
Comment (naushir) [reply encoder/libav_encoder.cpp:199]: True, I can get rid of the length constraint.

Comment (naushir) [reply encoder/libav_encoder.cpp:199]: Actually, this is the equivalent of ends_with(), so the constraint here is the start position of the search.

**Pull request merged**